### PR TITLE
API: map() on Index returns an Index, not array

### LINKED ISSUE #12766: API: Index.map should return Index rather than array
``` python
df2=pd.DataFrame([[0,1],[2,3]],columns=['c1','c2'],index=['i1','i2'])
df2.index.name='index'
df2.to_html()
```

``` html
<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>c1</th>\n      <th>c2</th>\n    </tr>\n    <tr>\n      <th>index</th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>i1</th>\n      <td>0</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>i2</th>\n      <td>2</td>\n      <td>3</td>\n    </tr>\n  </tbody>\n</table>
```

``` python
df2.index = df2.index.map(lambda x: x.upper())
df2.to_html()
```

``` html
<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>c1</th>\n      <th>c2</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>I1</th>\n      <td>0</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>I2</th>\n      <td>2</td>\n      <td>3</td>\n    </tr>\n  </tbody>\n</table>
```

I think `map` should accept `inplace`

```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 18.2
Cython: None
numpy: 1.10.4
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: None
patsy: None
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: None
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
```

### PR DESCRIPTION
**Pull request by nateyoder**
- [x] closes #12766
- [x] closes #12798 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

This is a follow on to #12798.

### CODE REVIEW

**Comment by nateyoder** [comment]
Note that the following changes result in some additional changes to the API which should potentially be addressed so I more than welcome suggestions.

- MultiIndex maps to Index: because mapping a function can reduce or increase the dimensionality of the index I am returning an Index from these operations rather than inspecting the data to determine whether the output should potentially be a MultiIndex
- Categorical and CategoricalIndex: When a categorical cannot be returned returns an Index of the appropriate type rather than an np.ndarray

**Comment by jreback** [comment]
@nateyoder use ``ensure_index`` to sniff for Index/Multi_index. You can simply return a ``CategoricalIndex`` (rather than a plain ``Categorical``).

This will need to be in 0.20.0

File: doc/source/whatsnew/v0.19.1.txt [hunk 1]
```diff
@@ -26,7 +26,29 @@ Performance Improvements
 - Improved performance in ``DataFrame.asof(where)`` when ``where`` is a scalar (:issue:`14461)
 
 
+.. _whatsnew_0191.api:
 
+API changes
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.1.txt]: move to 0.20.0

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -766,6 +766,26 @@ def test_sub(self):
         self.assertRaises(TypeError, lambda: idx - idx.tolist())
         self.assertRaises(TypeError, lambda: idx.tolist() - idx)
 
+    def test_map_identity_mapping(self):
+        for name, cur_index in self.indices.items():
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: can you add a comment with the appropriate github issue numbers (don't go crazy but where appropriate)

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -766,6 +766,26 @@ def test_sub(self):
...
+            if name == 'tuples':
+                expected = Index(cur_index.values, tupleize_cols=False)
+                self.assert_index_equal(expected, cur_index.map(lambda x: x))
+            else:
+                self.assert_index_equal(cur_index, cur_index.map(lambda x: x))
+
+    def test_map_that_returns_tuples_creates_index_not_multi_index(self):
+        boolean_index = tm.makeIntIndex(3).map(lambda x: (x, x == 1))
+        expected = Index([(0, False), (1, True), (2, False)],
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: use ``ensure_index`` and this will be a MI

**Comment by nateyoder** [comment]
@jreback I may have misunderstood what you were saying but I gave using _ensure_index a shot with the below code but it didn't seem to do anything.  I believe this may be because the map returns a np.ndarray rather than a list which would then be converted to a MultiIndex?  Should I go ahead an put a tolist in there despite the performance drawbacks? 

```
_ensure_index(Index(self._arrmap(self.values, mapper), **attributes))
```

**Comment by jreback** [comment]
hmm, actually it IS prob worth introspecting this and if it *is* ndim==1, then its no problem, if its is an array-like of tuples, then you can pass to ``MultiIndex.from_tuples`` directly.

File: pandas/core/categorical.py [hunk 4]
```diff
@@ -943,7 +943,7 @@ def map(self, mapper):
 
         Returns
         -------
-        applied : Categorical or np.ndarray.
+        applied : Categorical or Index.
```
Comment (nateyoder) [inline pandas/core/categorical.py]: 👍

**Comment by nateyoder** [comment]
Because the outputs of _arrmap were actually always contained in a 1d np.ndarray I actually checked for the presence of a tuple.  Let me know if you'd prefer something else.

File: pandas/indexes/base.py [hunk 5]
```diff
@@ -2436,9 +2436,19 @@ def map(self, mapper):
 
         Returns
         -------
-        applied : array
+        applied : Index
+            The output of the mapping function applied to the index.
+            If the function returns a tuple a
```
Comment (jorisvandenbossche) [inline pandas/indexes/base.py]: missing some content

File: pandas/tests/indexes/test_base.py [hunk 6]
```diff
@@ -766,6 +766,30 @@ def test_sub(self):
...
+        expected = MultiIndex.from_tuples([(0, False), (1, True), (2, False)])
+        self.assert_index_equal(boolean_index, expected)
+
+    def test_map_that_returns_a_length_one_tuple_creates_an_index(self):
+        # GH 12766
+        boolean_index = tm.makeIntIndex(3).map(lambda x: (x, ))
+        expected = Index([(0, ), (1, ), (2, )])
+        self.assert_index_equal(boolean_index, expected)
+
+    def test_map_that_reduces_multi_index_to_single_index_returns_index(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/test_base.py]: Can you shorten the test names a bit? 
(or else just include them in one single test, then you can add the long name/explanation as a comment)

File: doc/source/whatsnew/v0.20.0.txt [hunk 7]
```diff
@@ -38,6 +38,25 @@ Other enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+-  ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
+.. ipython:: python
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.20.0.txt]: Some sphinx sytax comments (sphinx is rather picky ..):

- blank line above `.. ipython ...`
- Can you also indent this line (and all lines below) with two spaces? (to be at the level of the content of the bullet point) -> so everything will be part of this bullet point

**Review by jorisvandenbossche — COMMENTED** [review]
For DatetimeIndex (and probably timedelta and period as well), the return value is still an array if the result is not a datetime anymore (eg with `.map(lambda x: 1)` or `.map(lambda x: x.hour)`

**Comment by nateyoder** [comment]
Great catch on the DatetimeIndex!  Sorry I missed that. I'll try to update the PR today.

**Comment by nateyoder** [comment]
Unfortunately when making this change I ended up causing two other tests (series.test_apply.TestSeriesApply.test_apply_datetimetz and series.test_apply.TestSeriesMap.test_map_datetimetz) to fail because they became np.int64 instead of np.int32.  Is this a significant failure that I should try to work around or should I change the dtype in the tests?

**Comment by nateyoder** [comment]
Just wanted to check and see if there were any additional actions you would like taken on this PR?  It appeared to me as though the AppVeyor failure was not due to my changes but am not familiar with AppVeyor.

File: doc/source/whatsnew/v0.20.0.txt [hunk 8]
```diff
@@ -38,8 +38,27 @@ Other enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. _whatsnew_0200.api:
+-  ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: make a new sub-section

File: pandas/tests/indexes/test_base.py [hunk 9]
```diff
@@ -766,6 +766,48 @@ def test_sub(self):
         self.assertRaises(TypeError, lambda: idx - idx.tolist())
         self.assertRaises(TypeError, lambda: idx.tolist() - idx)
 
+    def test_map_identity_mapping(self):
+        # GH 12766
+        for name, cur_index in self.indices.items():
+            self.assert_index_equal(cur_index, cur_index.map(lambda x: x))
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: I think we just use ``assert_index_equal`` (not the self), same routine, but more consistent

File: pandas/tests/indexes/test_category.py [hunk 10]
```diff
@@ -207,19 +207,20 @@ def test_map(self):
         ci = pd.CategoricalIndex(list('ABABC'), categories=list('CBA'),
                                  ordered=True)
         result = ci.map(lambda x: x.lower())
-        exp = pd.Categorical(list('ababc'), categories=list('cba'),
-                             ordered=True)
-        tm.assert_categorical_equal(result, exp)
+        exp = pd.CategoricalIndex(list('ababc'), categories=list('cba'),
```
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: show an example like this in the whatsnew as well (e.g. CategoryIndex.map -> CI rather than Category now)

File: pandas/tests/series/test_apply.py [hunk 11]
```diff
@@ -124,7 +124,7 @@ def test_apply_datetimetz(self):
 
         # change dtype
         result = s.apply(lambda x: x.hour)
-        exp = pd.Series(list(range(24)) + [0], name='XX', dtype=np.int32)
+        exp = pd.Series(list(range(24)) + [0], name='XX', dtype=np.int64)
```
Comment (jreback) [inline pandas/tests/series/test_apply.py]: hmm, I don't think this should have changed, these are normally ``int32``s

any idea? (also this might not be the same on windows)

File: pandas/tseries/base.py [hunk 12]
```diff
@@ -330,11 +330,16 @@ def _nat_new(self, box=True):
...
+            # Try to use this result if we can
+            if isinstance(result, np.ndarray):
+                self._shallow_copy(result)
+
+            if not isinstance(result, Index):
                 raise TypeError
             return result
         except Exception:
-            return _algos.arrmap_object(self.asobject.values, f)
+            return self.asobject.map(f)
```
Comment (jreback) [inline pandas/tseries/base.py]: this is better, using ``.values`` can change dtypes (esp timezones)

Comment (jreback) [inline pandas/tseries/base.py]: this might actually close another issue...

File: pandas/tseries/tests/test_converter.py [hunk 13]
```diff
@@ -104,7 +104,7 @@ def test_dateindex_conversion(self):
         for freq in ('B', 'L', 'S'):
             dateindex = tm.makeDateIndex(k=10, freq=freq)
             rs = self.dtc.convert(dateindex, None, None)
-            xp = converter.dates.date2num(dateindex._mpl_repr())
+            xp = Index(converter.dates.date2num(dateindex._mpl_repr()))
             tm.assert_almost_equal(rs, xp, decimals)
```
Comment (jreback) [inline pandas/tseries/tests/test_converter.py]: compare as index

File: pandas/tseries/tests/test_timedeltas.py [hunk 14]
```diff
@@ -1513,8 +1513,8 @@ def test_map(self):
 
         f = lambda x: x.days
         result = rng.map(f)
-        exp = np.array([f(x) for x in rng], dtype=np.int64)
```
Comment (jreback) [inline pandas/tseries/tests/test_timedeltas.py]: I might show some of these examples in the whatsnew as well (e.g. period/timedelta)

**Comment by jreback** [comment]
@nateyoder sorry for the delay. this looks pretty good. couple of comments. Pls have a look at the dtype changes w.r.t. ``.apply(lambda x: x.hour)``; these *should* be ``int32``.

@sinhrks thoughts

File: pandas/tests/indexes/test_base.py (hunk 9 above)
Comment (nateyoder) [inline pandas/tests/indexes/test_base.py]: Sounds good.  I changed the ones related to my commit but left the other ones in test_base.py alone.

File: pandas/tests/series/test_apply.py (hunk 11 above)
Comment (jorisvandenbossche) [inline pandas/tests/series/test_apply.py]: It's true it previously gave int32, but is there any reason for that? We almost always use int64 as the default integer size, and also currently Timestamp.hour gives you back int64 when you replace the `map` with an explicit loop:

```
In [20]: dtidx = pd.date_range(start='2012-01-01', periods=4)

In [29]: dtidx.map(lambda x: x.hour)
Out[29]: array([0, 0, 0, 0], dtype=int32)

In [30]: np.array([x.hour for x in dtidx])
Out[30]: array([0, 0, 0, 0])

In [31]: np.array([x.hour for x in dtidx]).dtype
Out[31]: dtype('int64')
```

Comment (jreback) [inline pandas/tests/series/test_apply.py]: this is actually an implementation detail (as they r stored as int32)

we can change but that should be separate PR

ideally just like to preserve here

**Comment by nateyoder** [comment]
Hi all.  Apologies for the delay on this after your thorough review.  I've updated the documentation and I think I've made almost all the changes you asked for.

However WRT the int32 vs int64 case I ran into a bit of an issue.  This seems to be happening because I need to return an index when map is applied to and index and current behavior is to force any numeric data type to the classes default rather than allowing the user to specify the dtype.  This can be seen here: https://github.com/pandas-dev/pandas/blob/master/pandas/indexes/numeric.py#L41 where the input dtype is completely ignored.  I can go ahead and make the change to use the input dtype but it would result in some potentially confusing situations where an Int64Index has a dtype of int32, int16, or even int8.  It would also of course impact other tests.  I've already started on these changes but before charging ahead thought I should probably see what you'd like to do.

Thanks!

**Comment by jreback** [comment]
@nateyoder construction of an Index with a particular kind of dtype is not allowed, e.g. we cannot have int32's in an Int64Index, only int64s. So you infer the best you can but the constructors will promote if needed; you can actually construct like this:

So it will respect the flavor but not the itemsize. This is no purpose.
```
In [1]: Index(np.array([1,2,3]),dtype='int32')
Out[1]: Int64Index([1, 2, 3], dtype='int64')

In [2]: Index(np.array([1,2,3]),dtype='float32')
Out[2]: Float64Index([1.0, 2.0, 3.0], dtype='float64')
```

**Comment by nateyoder** [comment]
Hi @jreback.  Sorry for my confusing message.  I understand that it is the current behavior of indices to ALWAYS return `int64` but if you would like to be able to return `int32` rather than `int64` types from map operations on Indices as you had said two weeks ago I believe that would have to change.  Would you like me to try to make those changes or are you alright with map operations always returning `int64` if an integer type is returned?  Apologies again for the confusion.

**Comment by jreback** [comment]
not sure what you are referring

**Comment by jorisvandenbossche** [comment]
@jreback I think the confusion is this: you ask that the return dtype *should* be int32 (as before: return type of an int32 array), but for an Index this can never be the case as even an int32 array will result in a int64 index.

So for `Index.map`, the resulting dtype *has* to change in this PR.

The question remains of course if it is possible to keep the dtype of `Series.map/apply`. But, this is already a bit inconsistent at the moment:

```
In [45]: dtidx = pd.date_range("2012-01-01", freq='H', periods=5)

In [46]: dtidx.map(lambda x: x.hour)
Out[46]: array([0, 1, 2, 3, 4], dtype=int32)

In [47]: dtidx2 = dtidx.tz_localize("Europe/Brussels")

In [48]: dtidx2.map(lambda x: x.hour)
Out[48]: array([0, 1, 2, 3, 4], dtype=int32)

In [49]: pd.Series(dtidx).map(lambda x: x.hour)
Out[49]: 
0    0
1    1
2    2
3    3
4    4
dtype: int64

In [50]: pd.Series(dtidx2).map(lambda x: x.hour)
Out[50]: 
0    0
1    1
2    2
3    3
4    4
dtype: int32
```

So for a series it gives in64 for datetimes, but int32 for timezone aware datetimes. So personally I wouldn't mind of all those case become int64 (for Index.map it will be in any case, only leaving tz aware Series to return int32)

**Comment by jreback** [comment]
@nateyoder @jorisvandenbossche response is good (I dont think this is a big deal that we r actually returning int32 from the datetime computations - can evaluate that on another issue)

I think my original comment was why we needed to change a test from int32 to int64; since we are always returning an Index which by definition have restricted dtypes the test needs to change

so ought to document this with an example of that (in whatsnew)

File: pandas/tests/indexes/test_category.py (hunk 10 above)
Comment (nateyoder) [inline pandas/tests/indexes/test_category.py]: 👍

**Comment by nateyoder** [comment]
Sounds good.  I updated the whatsnew and believe I have no addressed all of the comments.

File: doc/source/whatsnew/v0.20.0.txt [hunk 15]
```diff
@@ -38,8 +38,127 @@ Other enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. _whatsnew_0200.api:
 
+Map on Index types now return other Index types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+-  ``map`` on an ``Index`` now returns an ``Index``, not an array (:issue:`12766`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: array -> ``np.array``

File: doc/source/whatsnew/v0.20.0.txt [hunk 16]
```diff
@@ -38,8 +38,127 @@ Other enhancements
...
+
+  .. ipython:: python
+
+      idx.map(lambda x: x * 2)
+
+-  ``map`` on an ``Index`` or ``MultiIndex`` returns the appropriate type depending on output dimensionality
+
+  .. ipython:: python
+
+     mi = MultiIndex.from_tuples([(1, 2), (2, 4)])
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: put this with where you define the Index as well. You can then put the previous/new for Multi with Index

File: doc/source/whatsnew/v0.20.0.txt [hunk 17]
```diff
@@ -38,8 +38,127 @@ Other enhancements
...
+
+  .. ipython:: python
+
+      idx.map(lambda x: (x, x * 2))
+
+      mi.map(lambda x: x[0])
+
+
+-  ``map`` on an ``CategoricalIndex`` now returns a ``CategoricalIndex``, not a Categorical
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: you can skip this example (the CategoricalIndex) and below

File: pandas/indexes/base.py [hunk 18]
```diff
@@ -2436,9 +2435,21 @@ def map(self, mapper):
 
         Returns
         -------
-        applied : array
+        applied : Index
+            The output of the mapping function applied to the index.
+            If the function returns a tuple with more than one element
```
Comment (jreback) [inline pandas/indexes/base.py]: say the output Index type will be inferred

File: pandas/tseries/base.py [hunk 19]
```diff
@@ -330,11 +329,16 @@ def _nat_new(self, box=True):
...
         try:
             result = f(self)
-            if not isinstance(result, (np.ndarray, Index)):
+
+            # Try to use this result if we can
+            if isinstance(result, np.ndarray):
+                self._shallow_copy(result)
+
+            if not isinstance(result, Index):
                 raise TypeError
```
Comment (jreback) [inline pandas/tseries/base.py]: I *think* this is caught in an outer scope and so is not seen. but can you add an informative message to this ``TypeError``

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. just shorten up the whatsnew examples a bit. If you want refresh the docs a bit (with some examples), you can search where we use ``Index.map`` (I don't know if its even shown) in indexing.rst or advanced.rst. Can do this in a followup PR (or here if you want)

**Comment by jorisvandenbossche** [comment]
@nateyoder Could you update for the latest comments of @jreback? Would be nice to get this in!

File: doc/source/whatsnew/v0.20.0.txt (hunk 15 above)
Comment (nateyoder) [inline doc/source/whatsnew/v0.20.0.txt]: 👍

File: pandas/tests/indexes/test_category.py (hunk 10 above)
Comment (nateyoder) [inline pandas/tests/indexes/test_category.py]: Hi @jreback.  Just wanted to touchbase here since I added this example and then you asked me to remove in another comment below.  Did I add it in the wrong place or did you just decide it was a little overkill?  Thanks.

**Comment by nateyoder** [comment]
@jorisvandenbossche Sorry for the delay.  Let me know if you notice anything else that needs to be updated.

Thanks I've enjoyed my first contribution!

File: pandas/tests/indexes/test_category.py (hunk 10 above)
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: it's too long for the what's new; so need to pare it down

Comment (nateyoder) [inline pandas/tests/indexes/test_category.py]: 💯

File: doc/source/whatsnew/v0.20.0.txt [hunk 20]
```diff
@@ -91,8 +91,75 @@ Other enhancements
...
+      idx.map(lambda x: x * 2)
+
+      idx.map(lambda x: (x, x * 2))
+
+      mi.map(lambda x: x)
+
+      mi.map(lambda x: x[0])
+
+
+-  ``map`` on a Series withe datetime64 values may return int64 dtypes rather than int32
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: withe -> with

File: doc/source/whatsnew/v0.20.0.txt [hunk 21]
```diff
@@ -91,8 +91,75 @@ Other enhancements
...
+
+  .. ipython:: python
+
+    s = Series(date_range('2011-01-02T00:00', '2011-01-02T02:00', freq='H').tz_localize('Asia/Tokyo'))
+    s
+
+  Previous Behavior:
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: can you flip the order of these (previous, then last)

**Comment by jreback** [comment]
lgtm. just a small doc change.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

great effort!

**Comment by jorisvandenbossche** [comment]
@nateyoder Thanks a lot!
